openmeter/entitlement/metered/grant_owner_adapter.go

```diff
@@ -107,19 +108,52 @@ func (e *entitlementGrantOwner) GetStartOfMeasurement(ctx context.Context, owner
 	return metered.MeasureUsageFrom, nil
 }
 
+// The current usage period start time is either the current period start time, or if this is the first period then the start of measurement
 func (e *entitlementGrantOwner) GetUsagePeriodStartAt(ctx context.Context, owner grant.NamespacedOwner, at time.Time) (time.Time, error) {
-	// If this is the first period then return start of measurement, otherwise calculate based on anchor.
-	// To know if this is the first period check if usage has been reset.
+	ent, err := e.entitlementRepo.GetEntitlement(ctx, owner.NamespacedID())
```

Comment:
Maybe not for this specific PR, but these should be refactored to not to fetch the entitlement multiple times during engine building. E.g. this one is called 2-3 times during a balance query, having to refetch entitlements multiple times from the db.